ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll

```diff
@@ -0,0 +1,423 @@
+  GraphqlFieldDefinitionMethodCall getFieldDefinition() { result = fieldDefinition }
+
+  /** Gets the name of the field which this is an argument for */
+  string getFieldName() { result = this.getFieldDefinition().getFieldName() }
+
+  /** Gets the name of the argument (i.e. the first argument to this `argument` method call) */
+  string getArgumentName() { result = this.getArgument(0).(SymbolLiteral).getValueText() }
+}
+
+/**
+ * A `Method` which represents an instance method which is the resolver method for a
+ * GraphQL `field`.
+ *
+ * For example:
+ *
+ * ```rb
+ * class Types::User < GraphQL::Schema::Object
+ *   field :email, String
+ *   field :friends, [Types::User] do
+ *     argument :starts_with, String, "Show only friends matching the given prefix"
+ *   end
+ *
+ *   # GraphqlFieldResolutionMethod
+ *   def friends(starts_with:)
+ *     object.friends.where("name like #{starts_with}")
+ *   end
+ * end
+ * ```
+ *
+ * or:
+ *
+ * ```rb
+ * class Types::User < GraphQL::Schema::Object
+ *   field :email, String
+ *   field :friends, [Types::User] do
+ *     argument :starts_with, String, "Show only friends matching the given prefix",
+ *       resolver_method: :my_custom_method, extras: [:graphql_name]
+ *   end
+ *
+ *   # GraphqlFieldResolutionMethod
+ *   def my_custom_method(**args)
+ *     puts args[:graphql_name] # for debugging
+ *     object.friends.where("name like #{args[:starts_with]}")
+ *   end
+ * end
+ * ```
+ */
+class GraphqlFieldResolutionMethod extends Method, HTTP::Server::RequestHandler::Range {
+  private GraphqlSchemaObjectClass schemaObjectClass;
+
+  GraphqlFieldResolutionMethod() {
+    this.getEnclosingModule() = schemaObjectClass and
+    exists(GraphqlFieldDefinitionMethodCall defn |
+      // field :foo, resolver_method: :custom_method
+      // def custom_method(...)
+      defn.getKeywordArgument("resolver_method").(SymbolLiteral).getValueText() = this.getName()
+      or
+      // field :foo
+      // def foo(...)
+      not exists(defn.getKeywordArgument("resolver_method").(SymbolLiteral)) and
+      defn.getFieldName() = this.getName()
+    )
+  }
+
+  /** Gets the method call which is the definition of the field corresponding to this resolver method. */
+  GraphqlFieldDefinitionMethodCall getDefinition() {
+    result.getKeywordArgument("resolver_method").(SymbolLiteral).getValueText() = this.getName()
+    or
+    not exists(result.getKeywordArgument("resolver_method").(SymbolLiteral)) and
+    result.getFieldName() = this.getName()
+  }
+
+  // check for a named argument the same name as a defined argument for this field
+  override Parameter getARoutedParameter() {
+    result = this.getAParameter() and
+    exists(GraphqlFieldArgumentDefinitionMethodCall argDefn |
+      argDefn.getEnclosingCallable() = this.getDefinition().getBlock() and
+      (
+        result.(NamedParameter).hasName(argDefn.getArgumentName())
```

Comment:
From looking at the graphql-ruby [docs](https://graphql-ruby.org/fields/arguments), it looks like Graphql arguments are always passed as keyword arguments, so should we be using `KeywordParameter` here?

If so, that means we can drop line 409 as `KeywordParameter` and `HashSplatParameter` are disjoint.

Reply:
yep, makes sense.